airflow/models.py

```diff
@@ -2777,3 +2766,150 @@ class ImportError(Base):
     timestamp = Column(DateTime)
     filename = Column(String(1024))
     stacktrace = Column(Text)
+
+
+class Trigger:
+    """
+    Triggers represent a dependency of an operator on a range of schedules of
+    another operator.
+    * Trigger(t) depends on the state of t of the same schedule
+    * Trigger(t, 1) depends on the state of t of the previous schedule
+    * Trigger(t, n) depends on the state of t of the n previous schedules
+      (not including the current schedule)
+
+    :param task: the depended upon task
+    :type task: TaskInstance
+    :param past_executions: The number of past executions that the
+    dependant depends on. If it is left to 0, the Trigger is aimed at
+    the same schedule as the dependant. If it is set to a value > 0,
+    the Trigger will look at the corresponding past executions to
+    evaluate the Trigger.
+    :type past_executions: non negative integer
+    """
+
+    def __init__(self, task, past_executions=0):
+        self.dag_id = task.dag_id
+        self.task_id = task.task_id
+        self.task = task
+
+        assert past_executions >= 0, "past_executions parameter must be >= 0"
+        self.past_executions = past_executions
+
+    def states(self, execution_date, main_session=None):
+        """
+        Returns all the states of depended upon task instances.
+
+        :param execution_date: the date on which the trigger should be evaluated
+        :type execution_date: datetime
+        :return: the states of all depended upon task instances
+        :rtype: list of airflow.utils.State
+        """
+        TI = TaskInstance
+
+        first_date = self._compute_first_date(execution_date)
+        last_date = self._compute_last_date(execution_date)
+
+        if last_date < first_date:
+            return []
+
+        session = main_session or settings.Session()
+
+        query = session.query(TI.execution_date, TI.state).filter(
+            TI.dag_id == self.dag_id,
+            TI.task_id == self.task_id,
+            TI.execution_date >= first_date,
+            TI.execution_date <= last_date,
+        )
+
+        if not main_session:
+            session.commit()
+            session.close()
+
+        sparse_results = dict(query.all())
+        schedules = self.task.dag.date_range(start_date=first_date,
+                                             end_date=last_date)
+
+        # Filling states with Nones when the TI isn't in the DB yet
+        states = []
+        for schedule in schedules:
```

Comment:
putting the construction of `states` in a separate function would make it unit-testable